source/common/http/http2/metadata_decoder.cc

```diff
@@ -0,0 +1,102 @@
+#include "common/http/http2/metadata_decoder.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Http {
+namespace Http2 {
+
+MetadataDecoder::MetadataDecoder(uint64_t stream_id) : stream_id_(stream_id) {
+  int rv = nghttp2_hd_inflate_new(&inflater_);
+  ASSERT(rv == 0);
+}
+
+MetadataDecoder::~MetadataDecoder() { nghttp2_hd_inflate_del(inflater_); }
+
+void MetadataDecoder::receiveMetadata(const uint8_t* data, size_t len) {
+  if (data == nullptr || len == 0) {
+    ENVOY_LOG(error, "No payload for the decoder to receive.");
+    return;
+  }
+
+  Buffer::RawSlice iovec;
+  payload_.reserve(len, &iovec, 1);
+  ASSERT(iovec.len_ >= len);
+  iovec.len_ = len;
+  memcpy(iovec.mem_, data, len);
+  payload_.commit(&iovec, 1);
+}
+
+bool MetadataDecoder::OnMetadataFrameComplete(bool end_metadata) {
+  DecodeMetadataPayloadUsingNghttp2(end_metadata);
+
+  if (end_metadata) {
+    if (callback_ != nullptr) {
+      callback_(metadata_map_);
+      metadata_map_.clear();
+    } else {
+      metadata_map_list_.emplace_back(std::move(metadata_map_));
```

Comment:
I think comments would help but I'm not quite clear why we're creating a metadata_map_ per frame rather than just have all the name value pairs across frames going into one map

Reply:
METADATA can cross multiple frames depends on the max frame size. If it does, the last frame is indicated by END_METADATA flag. metadata_map_ hold a set of METADATA received until END_METADATA flag. 
I can put all sets of METADATA into one map. But Shinkanzen METADATA separates different set into different maps. I can remove metadata_map_list_ for now. If Biren confirms it is necessary, I can add it back later. What do you think?

Reply:
Ah, I think I may be confused about naming.

Based on your explanation (thanks!) this isn't so much called when a metadata frame is complete as it's called when a logical unit of metadata is complete, right?  If we have 4 HTTP2 frames  [metadata a][metadata a''][metadata a''' 0x4][metadata b 0x4] this function would be called two times (at end-of-a and end-of-b) not 4 times?  If so I think we simply need to rename to onMetadataDataComplete()

Reply:
Sorry about the confusion! Whenever a Metadata frame is received, we will feed the payload to the decoder to do the incremental decoding. When a logical unit of metadata is complete, the registered callback will be called. So, in the example, onMetadataFrameComplete() will be called 4 times, and the callback will be called twice.

Reply:
Oh right, that totally makes sense - I blame insufficient caffeine. OK, with one more visual pass I'm totally fine with this section - sorry for all the confusion. :-)